### PR TITLE
skip hpa when rolling update is in progress

### LINKED ISSUE #72775: pod hpa would create extra pods during deployment rolling update when there is no load at all during the rolling upgrade
**What happened**:
I have deployment with following upgrade strategy

>     strategy:
>       rollingUpdate:
>         maxSurge: 1
>         maxUnavailable: 0
>       type: RollingUpdate
> 

and hpa:

>  spec:
>     maxReplicas: 3
>     minReplicas: 1
>     scaleTargetRef:
>       apiVersion: extensions/v1beta1
>       kind: Deployment
>       name: xxxx
>     targetCPUUtilizationPercentage: 80

When sending patch command to update container image, (current replicas is 1),  following happended
1. new pod created and getting ready
2. after the new pod ready, the old pod is terminating.
3. hpa make a judgement call to set desired replica to 2
4. after a while, hpa readjust to desired replica back to 1

**What you expected to happen**:
1. new pod created and getting ready
2. after the new pod ready, the old pod is terminating.

**How to reproduce it (as minimally and precisely as possible)**:
1. create a deployment with replica set to 1, maxSurge 1, maxUnavailable: 0
2. create a hpa targeting the deployment with replica 1 from 3, 80% cpu target
3. change container image via kubectl edit or other approach
4. keep checking pod count changes

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by zq-david-wang**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
 /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
skip hpa rescale if pods have multiple owner, which would happen when deployment rolling update is in progress
**Which issue(s) this PR fixes**:

Fixes #72775 

**Special notes for your reviewer**:
Current hpa code make calculation without distinguishing whether all replicas belong to current resource revision, but write back the result into current revision. e.g. during rolling update for a deployment with replica tuned to 8, there could be 8 old replica for old revision and 4 replica for new revision, hpa would collect metrics for all those 12 replicas, if no scale needed, hpa would not change the desired replica  but hpa would store 12 into its normalized table. and when rolling update finished, replicas becomes 8, hpa calculate next replica should also be 8. then when hpa check against its normalized table, it would change replica count to 12 and make unnecessary scale up.

This code change would skip hpa if  something else is in progress by checking whether all pods have same owner id. 
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/podautoscaler/replica_calculator.go [hunk 1]
```diff
@@ -340,13 +349,26 @@ func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(currentReplicas int3
...
+	var podOwnerID types.UID
 	for _, pod := range pods {
 		if pod.DeletionTimestamp != nil || pod.Status.Phase == v1.PodFailed {
+			// remove pod from metrics
+			delete(metrics, pod.Name)
 			continue
 		}
+		if len(pod.OwnerReferences) > 0 {
+			if podOwnerID != pod.OwnerReferences[0].UID {
+				if podOwnerID != types.UID("") {
```
Comment (Rajat-0) [inline pkg/controller/podautoscaler/replica_calculator.go]: why we need this check?

Comment (zq-david-wang) [reply pkg/controller/podautoscaler/replica_calculator.go]: Sorry for the late response, i was on vacation.
I am not sure whether you were asking about the if statements or about why I want to check whether pods have multiple owners.
For those if statements, I want to check whether 'podOwnerID' is assigned a valid uid yet. (by default, its avlue is ""), if it was already assigned a valid owner uid, but is different from current pod's owner uid, then we can conclude that the pods have multiple owners.

I think counting owners is one solution to distinguish those situations, e.g. rolling upgrade,  that I do not want hpa to take actions.

Comment (Rajat-0) [reply pkg/controller/podautoscaler/replica_calculator.go]: Thanks for the clarification, you cleared my doubt.

**Comment by mwielgus** [comment]
Deployment upgrade can take quite a long time. During that time HPA should be operational. Stopping it completely is not the right approach.

**Comment by zq-david-wang** [comment]
@mwielgus Agree that there are situations one would want to speed up upgrade process; but I do not agree that this should be controlled by hpa. During upgrade, deployment controller has maxserge/maxunavailable to control replicas during upgrade which can control min/max pods, (but they are static configuration.)

I think It is quite dangerous when multiple modules try to control replicas at the same time.
If upgrade speed should be adjust dynamically, deployment controller should take it over.

**Comment by mwielgus** [comment]
@zq-david-wang HPA may have some extra code to appropriately handle actuations when rolling upgrade is running. Stopping HPA altogether is IMHO not the best option.

**Comment by zq-david-wang** [comment]
@mwielgus  I am not convinced that hpa code handle upgrade process correctly, that is the reason I file an issue #72775 and this PR, but code changes fast those days, maybe hpa code do have changed since I filed the bug.  
Anyway, it is ok for me that you guys do not want  this PR,  as long as someone takes a look at issue #72775

Close this PR since hpa would have better code logic to handle upgrade.

**Pull request closed without merging**